### PR TITLE
fix: Dashboard access when DASHBOARD_RBAC is disabled

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
This PR fixes the security manager to also check for dashboard access when `DASHBOARD_RBAC` is disabled. Previously, if `DASHBOARD_RBAC` was disabled, this piece of code was skipped:

```
can_access = (
    is_user_admin()
    or is_owner(dashboard, g.user)
    or (dashboard.published and has_rbac_access)
)
```

I also changed some endpoints to emit a 403 when an access error occurs.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: tests/integration_tests/dashboards/api_tests.py [hunk 1]
```diff
@@ -395,7 +395,7 @@ def test_get_dashboard_no_data_access(self):
         self.login(username="gamma")
         uri = f"api/v1/dashboard/{dashboard.id}"
         rv = self.client.get(uri)
-        self.assertEqual(rv.status_code, 200)
+        self.assertEqual(rv.status_code, 403)
```
Comment (betodealmeida) [inline tests/integration_tests/dashboards/api_tests.py:398]: Nit (I know this was already using `self.assertEqual`, but might as well update it):

```suggestion
        assert rv.status_code == 403
```

**Review by betodealmeida — APPROVED** [review]
This looks great, thanks for fixing!

**Comment by michael-s-molina** [comment]
@villebro @amitmiran137 these tests in `dashboard/api_tests.py` are checking for specific conditions:

```
@pytest.mark.usefixtures("load_world_bank_dashboard_with_slices")
def test_get_draft_dashboard_datasets(self):
    """
    All users should have access to dashboards without roles
    """
    self.login(username="gamma")
    uri = "api/v1/dashboard/world_health/datasets"
    response = self.get_assert_metric(uri, "get_datasets")
    self.assertEqual(response.status_code, 200)

@pytest.mark.usefixtures("create_dashboards")
def test_get_draft_dashboard_charts(self):
    """
    All users should have access to draft dashboards without roles
    """
    self.login(username="gamma")
    dashboard = self.dashboards[0]
    uri = f"api/v1/dashboard/{dashboard.id}/charts"
    response = self.get_assert_metric(uri, "get_charts")
    assert response.status_code == 200
```

I added the following line giving access to any user if the dashboard is not published and has no roles. 

https://github.com/apache/superset/blob/893595ec1a5e9d2d7b27ac7a7c7cf053905cf124/superset/security/manager.py#L1189

Can you confirm that this is the expected behavior?

**Review by amitmiran137 — APPROVED** [review]
The original intention here was to keep the old behaviour before DASHBOARD_RBAC era where there was no enforcement to dashboard access at all 

But I agree that access should be enforced.

There was another idea by @dpgaspar 
That We might want to consider switching to 404 this would prevent malicious API calls from finding out which dashboard IDs exist and which are not.

Altough personaly I prefer 403

**Comment by villebro** [comment]
> The original intention here was to keep the old behaviour before DASHBOARD_RBAC era where there was no enforcement to dashboard access at all
> 
> But I agree that access should be enforced.
> 
> There was another idea by @dpgaspar That We might want to consider switching to 404 this would prevent malicious API calls from finding out which dashboard IDs exist and which are not.
> 
> Altough personaly I prefer 403

Right, I remember we were pretty prudent during review in making sure the behavior was unchanged when the feature flag was disabled. Regarding 403/404 , I'm also more a 403 type of guy for this sort of stuff, but I'm ok with 404 if that's the security pattern we've chosen.

**Comment by villebro** [comment]
I tagged @john-bodley and @etr2460 for reviews - can you think of a reason why dashboard access should not be enforced? The `raise_for_dashboard_access` method was added in #12875 - previously this didn't exist at all. To me it seems this access should be enforced.

**Review by villebro — APPROVED** [review]
LGTM

File: superset/dashboards/api.py [hunk 2]
```diff
@@ -327,6 +332,8 @@ def get_datasets(self, id_or_slug: str) -> Response:
               $ref: '#/components/responses/400'
             401:
               $ref: '#/components/responses/401'
+            403:
+              $ref: '#/components/responses/403'
```
Comment (dpgaspar) [inline superset/dashboards/api.py:336]: I'm fine with returning 403, although it's more coherent with the current setup that we return 404, note that the dashboard filter is serving has security so we only "show" resources that are available to the user. Also, exposing less detailed info about why a resource is not available for access to a user the better.

RFC is not 100% clear regarding this:
https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.3
https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.4
it's acceptable to return 404 if we don't want to disclose that the dashboard exists but it's access is forbidden to the user.

**Review by dpgaspar — APPROVED** [review]
Looks good, personally I think that this is getting hard to follow, security filters set here: https://github.com/apache/superset/blob/master/superset/dashboards/filters.py#L72 should probably use the same code path for `raise_for_dashboard_access`

File: superset/dashboards/api.py (hunk 2 above)
Comment (michael-s-molina) [reply superset/dashboards/api.py:336]: @dpgaspar We have other endpoints also returning 403. I'll bring this discussion to our meeting today and if we decide to return 404 I'll open another PR fixing all endpoints.
